amaranth/hdl/_mem.py

```diff
@@ -1,33 +1,158 @@
 import operator
 from collections import OrderedDict
+from collections.abc import MutableSequence
 
 from .. import tracer
 from ._ast import *
 from ._ir import Elaboratable, Fragment
 from ..utils import ceil_log2
-from .._utils import deprecated
+from .._utils import deprecated, final
 
 
-__all__ = ["Memory", "ReadPort", "WritePort", "DummyPort"]
+__all__ = ["MemoryData", "Memory", "ReadPort", "WritePort", "DummyPort"]
 
 
-class MemoryIdentity: pass
+@final
+class FrozenError(Exception):
+    """This exception is raised when ports are added to a :class:`Memory` or its
+    :attr:`~Memory.init` attribute is changed after it has been elaborated once.
+    """
+
+
+class MemoryData:
+    class Init(MutableSequence):
+        """Memory initialization data.
+
+        This is a special container used only for the :attr:`Memory.init` attribute. It is similar
```

Comment:
This should not refer to `Memory.init` since that's from `lib`, not `hdl`. Maybe "This is a special container used only for initial contents of memories."?